src/shogun/base/progress.h

```diff
@@ -0,0 +1,434 @@
+/*
+* Written (W) 2017 Giovanni De Toni
+*/
+
+#ifndef __SG_PROGRESS_H__
+#define __SG_PROGRESS_H__
+
+#include <iterator>
+#include <memory>
+#include <string>
+
+#include <shogun/base/init.h>
+#include <shogun/base/range.h>
+#include <shogun/io/SGIO.h>
+#include <shogun/lib/Time.h>
+#include <shogun/mathematics/Math.h>
+
+#if WIN32
+#include <windows.h>
+#else
+#include <sys/ioctl.h>
+#include <unistd.h>
+#endif
+
+namespace shogun
+{
+
+	/** Possible print modes */
+	enum SG_PRG_MODE
+	{
+		ASCII,
+		UTF8
+	};
+
+	/** @class Printer that display the progress bar
+	*
+	*/
+	class ProgressPrinter
+	{
+	public:
+		/** Creates a ProgressPrinter instance given an SGIO object
+		*
+		* @param    io  SGIO object
+		*/
+		ProgressPrinter(
+		    const SGIO& io, float64_t max_value, float64_t min_value)
+		    : m_io(io), m_max_value(max_value), m_min_value(min_value),
+		      m_prefix("PROGRESS: "), m_mode(UTF8), m_last_progress(0),
+		      m_last_progress_time(0), m_progress_start_time(0)
+		{
+		}
+		ProgressPrinter(
+		    const SGIO& io, float64_t max_value, float64_t min_value,
+		    const std::string& prefix)
+		    : m_io(io), m_max_value(max_value), m_min_value(min_value),
+		      m_prefix(prefix), m_mode(UTF8), m_last_progress(0),
+		      m_last_progress_time(0), m_progress_start_time(0)
+		{
+		}
+		ProgressPrinter(
+		    const SGIO& io, float64_t max_value, float64_t min_value,
+		    const SG_PRG_MODE mode)
+		    : m_io(io), m_max_value(max_value), m_min_value(min_value),
+		      m_prefix("PROGRESS: "), m_mode(mode), m_last_progress(0),
+		      m_last_progress_time(0), m_progress_start_time(0)
+		{
+		}
+		ProgressPrinter(
+		    const SGIO& io, float64_t max_value, float64_t min_value,
+		    const std::string& prefix, const SG_PRG_MODE mode)
+		    : m_io(io), m_max_value(max_value), m_min_value(min_value),
+		      m_prefix(prefix), m_mode(mode), m_last_progress(0),
+		      m_last_progress_time(0), m_progress_start_time(0)
+		{
+		}
+		~ProgressPrinter()
+		{
+		}
+
+		/** Print the progress bar */
+		void print_progress(float64_t current_value) const
+		{
+			// Check if the progress was enabled
+			if (!m_io.get_show_progress())
+				return;
+
+			if (m_max_value <= m_min_value)
+				return;
+
+			// Check for terminal dimension. This is for provide
+			// a minimal resize functionality.
+			set_screen_size();
+
+			float64_t difference = m_max_value - m_min_value, v = -1,
+			          estimate = 0, total_estimate = 0;
+			float64_t size_chunk = -1;
+
+			// Check if we have enough space to show the progress bar
+			// Use only a fraction of it to account for the size of the
+			// time displayed (decimals and integer).
+			int32_t progress_bar_space =
+			    (m_columns_num - 50 - m_prefix.length()) * 0.9;
+
+			// TODO: this guy here brokes testing
+			// REQUIRE(
```

Comment:
It is calculated dynamically. There is a formula above which will calculate how much space we will use to print the progress bar. Sometimes the result can be 0 (or less), and that is a case in which we cannot print the progress bar.